File: src/code/trie/trie.go
type TrieNode struct {
	children [26]*TrieNode
	isEnd    bool
}

func NewTrieNode() *TrieNode {
	return &TrieNode{
		children: [26]*TrieNode{},
		isEnd:    false,
	}
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{root: NewTrieNode()}
}

func (trie *Trie) Insert(word string) {
	curr := trie.root
	for _, c := range word {
		index := c - 'a'
		if curr.children[index] == nil {
			curr.children[index] = NewTrieNode()
		}
		curr = curr.children[index]
	}
	curr.isEnd = true
}

func (trie *Trie) Search(word string) bool {
	curr := trie.root
	for _, c := range word {
		index := c - 'a'
		if curr.children[index] == nil {
			return false
		}
		curr = curr.children[index]
	}
	return curr.isEnd
}

func (trie *Trie) StartsWith(prefix string) bool {
	curr := trie.root
	for _, c := range prefix {
		index := c - 'a'
		if curr.children[index] == nil {
			return false
		}
		curr = curr.children[index]
	}
	return true
}

File: src/lang/python/trie.py
class TrieNode:
    __slots__ = ("children", "is_end")

    def __init__(self):
        # 固定 26 个字母位置
        self.children = [None] * 26
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            index = ord(ch) - ord('a')
            if node.children[index] is None:
                node.children[index] = TrieNode()
            node = node.children[index]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            index = ord(ch) - ord('a')
            if node.children[index] is None:
                return False
            node = node.children[index]
        return node.is_end

    def starts_with(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            index = ord(ch) - ord('a')
            if node.children[index] is None:
                return False
            node = node.children[index]
        return True
